### ISSUE TITLE
eslint using `@typescript-eslint/parser` looks for `tsconfig.json` in the wrong directory.

### ISSUE THREAD

**Issue by timfee** [zed-industries/community#1598]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

See https://github.com/timfee/zed-eslint

1. `pnpm create next-app` with default options.
2. change the parser to `@typescript-eslint/parser`, for example: 
```json
{
  "extends": [
    "next/core-web-vitals",
    "plugin:@typescript-eslint/recommended"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2021,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  }
}
```
3. observe that Zed reports an eslint error that `/users/timfee/sites/zed-eslint/app/tsconfig.json` cannot be read (it shouldn't be looking in that folder; it should be looking one level up, in the root).

Note: If I add `./test.ts` to the root folder of the project, it doesn't complain. \
Note 2: It tells me `./tailwind.config.js` was included by way of parserOptions, but it isn't.

Also, it seems to be ignoring case (not sure if that's an issue; but on Mac, my `Users` and `Sites` folder have initial caps).

### Environment

Zed: v0.88.6 (stable)
OS: macOS 13.4.0
Memory: 64 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![Screen Shot 2023-06-01 at 17 52 12](https://github.com/zed-industries/community/assets/3246342/55cc250a-6ae1-43b8-98b8-b897d97c9107)

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

[Zed.log](https://github.com/zed-industries/community/files/11630903/Zed.log)

**Comment by hovsater** [comment]
Since you've not generated a `src` directory, you should include a `root: true` in your ESLint configuration. That should solve your problem. So, change your `.eslintrc.json` like so:

```json
{
  "extends": [
    "next/core-web-vitals",
    "plugin:@typescript-eslint/recommended"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2021,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "root": true,
}
```

**Comment by timfee** [comment]
I had tried that before (it didn't work). I retried (see example repro), and I still get the same error.

I don't get this error (even without `root: true`) in VSCode.

Thanks!

**Comment by hovsater** [comment]
I can’t reproduce this. Make sure you restart your workspace upon changing your ESLint configuration. Does that help?

**Comment by timfee** [comment]
Thanks - sorry this is being such a weird repro! 

I restarted Zed completely each time. This is the latest:

![Screen Shot 2023-06-05 at 12 42 20](https://github.com/zed-industries/community/assets/3246342/a290f2a8-20a1-4fcd-b24b-6f48e08c6e71)

**Comment by hovsater** [comment]
No worries. 🙂 I’ll try again, and report back.

**Comment by timfee** [comment]
btw, this may be just a UI quirk - but the lowercasing catches me off guard:

`/Users/timfee/Sites/zed-eslint`

If it's WAI from a UI perspective, that's totally fine -- and not sure if this is causing any of the other issues. I don't have any non-standard HD partitioning (e.g. just case-insensitive APFS)

**Comment by timfee** [comment]
> No worries. 🙂 I’ll try again, and report back.

Let me know if I can help repro anything or screenshare / whatever the cool kids are doing. :-)  FWIW, I'm using `pnpm` without any global packages

**Comment by hovsater** [comment]
Alright. I've looked into this some more. The reason I could not re-produce this earlier was due to the fact that I didn't use the exact same ESLint configuration as you did. The project is the following option:

```json
{
    "parserOptions": {
        "project": "./tsconfig.json"
    }
}
```

It will try to resolve your `tsconfig.json` file relative to whatever is current file. The recommend approach here is to rename your `.eslintrc.json` to `.eslintrc.cjs` and then use the following configuration instead:

```javascript
{
    parserOptions: {
        project: true,
        tsconfigRootDir: __dirname,
    }
}
```

This is only relevant if you also extend your configuration with `plugin:@typescript-eslint/recommended-requiring-type-checking`. If you're not interested in [linting with type information](https://typescript-eslint.io/linting/typed-linting), you could just drop `project` line from `parserOptions` and it should now work as expected.

**Comment by hovsater** [comment]
Glad I could help! 🙂

**Comment by ryndshn** [comment]
i was getting this too and fixed it by removing the following from `.eslintrc.js`...
```parserOptions: {
        project: "./tsconfig.json"
    },
```
@hovsater thanks so much for the assistance. any idea why it works this way though? the functionality seems different from VSCode and frankly i'm afraid to modify my companies eslint config 😅

**Comment by hovsater** [comment]
It looks like the `tsconfig.json` is being resolved relative to the current working directory (instead of relative to the ESLint configuration). This seem to be [by design](https://github.com/typescript-eslint/typescript-eslint/issues/540). The solution would be to set [tsconfigRootDir](https://typescript-eslint.io/architecture/parser/#tsconfigrootdir) to `__dirname`.

Why does this differ from VSCode? I can't confidently answer that, but I'm guessing that the current working directory is the project root in VSCode whenever ESLint runs.

**Comment by thetutlage** [comment]
Also, the `cjs` workaround won't work for people storing their ESLint config inside `package.json` file. I do it all the time to keep the root of my project clean

**Comment by afernandr** [comment]
In my case, I have apps that contains several tsconfig files in the same repo. For that reason, I absolutely need to specify the file with the `project` field in `parserOptions`. I tried the solution from [here](https://github.com/zed-industries/community/issues/1598#issuecomment-1577428750) and it works in Zed. However, the new config breaks in VSCode, so I can't use Zed while my colleagues working on the same repo use VSCode.

This is the only problem that is keeping me from adopting Zed for all my projects.

**Comment by SomeoneToIgnore** [comment]
Seems to resemble the https://github.com/zed-industries/community/issues/2124 issue which got fixed in Zed 0.114.2 and Zed Preview 0.115.1.
So hopefully tsconfigs are now detected properly.
